### ISSUE TITLE
Tap on the screen

### ISSUE THREAD

**Issue by arajar** [Unity-Technologies/InputSystem#194]
Hi!
Is there a way to fake the input of the player and send tap / move commands to the system?

I'm trying to achieve this by using the Mouse device and Warping the position (based on the code of Devices_CanWarpMousePosition from the CoreTests_Devices.cs) but so far It's not working for me, as I don't see the mouse cursor moving around the screen.

Also, can we "click" on the screen by code?
Thanks a lot.

**Comment by Rene-Damm** [comment]
You can. Though there's gotchas. Also depends a bit on whether you need the actual cursor in the system to change or not. ATM on Windows supports cursor warping. On other platforms, it has no affect. Some will get implementations in the backend but not all platforms will support the feature.

For changing the state of arbitrary devices, the system allows sending events against any device you want. So if you want to alter the state of the mouse, all you need to do is make up some events and send them. For example, to move the left mouse button to down state, one could do

```
var mouse = Mouse.current;
InputEventPtr eventPtr;
using (StateEvent.From(mouse, out eventPtr))
{
    eventPtr.time = mouse.lastUpdateTime;
    mouse.leftButton.WriteValueInto(eventPtr, 1);
    InputSystem.QueueEvent(eventPtr);
    // Next update will pick up the event.
}
```

The gotcha for doing this is that the events you're sending may interfere with the events coming in from the backend (this is why I'm setting the event time to `lastUpdateTime`; basically it injects events into the event stream without moving time on the device forward).

**Comment by arajar** [comment]
Thanks a lot, that was really helpful!

Edit:
I am currently testing the system and so far it works, I created a small script with the following:

		if (Input.GetKeyDown(KeyCode.S))
		{
			Debug.Log("GetKeyDown");
			var mouse = Mouse.current;
			InputEventPtr eventPtr;
			using (StateEvent.From(mouse, out eventPtr))
			{
				eventPtr.time = mouse.lastUpdateTime;
				mouse.leftButton.WriteValueInto(eventPtr, 1);
				InputSystem.QueueEvent(eventPtr);
			}
		}

		if (Input.GetKeyUp(KeyCode.S))
		{
			Debug.Log("GetKeyUp");
			var mouse = Mouse.current;
			InputEventPtr eventPtr;
			using (StateEvent.From(mouse, out eventPtr))
			{
				eventPtr.time = mouse.lastUpdateTime;
				mouse.leftButton.WriteValueInto(eventPtr, 0);
				InputSystem.QueueEvent(eventPtr);
			}
		}

Checking the Input Debugger window the leftButtton and phase are correct (the do the same as when I click an actual mouse button), but the click per se is not happening. To test it I have several objects in the scene with this small script 
	`public void OnMouseDown()
	{
		Debug.Log("AAAAAAAAAAAAAAAAAAAAAAAAAAA");
	}`

But that log is only appearing in the console if I click with the mouse on the object, not with the script testing the Input System. Should that work?

**Comment by Rene-Damm** [comment]
Ah yes, I should have thought a little farther ahead :) ATM there is no connection from the new input system back to MonoBehaviour callbacks such as `OnMouseDown`. And no UI integration either (meaning that the UI won't respond to the changes you make to the mouse either).

UI integration we're going to work on soon. The MonoBehaviour stuff is much trickier as that happens all natively inside the Unity runtime. We will have to look at that as part of working towards deprecating the old input system but it'll be a while before we'll get to that.

So, in short, to amend my initial answer... you can but only with limited effects on Unity as a whole at this point.

**Comment by arajar** [comment]
Oh, I see. So far it's ok for me, I can more or less record and replicate all the taps on the screen made by a user, combining the new Input system with the old one (I'm wrapping all that stuff in a class, kind of like TouchKit). 
It's not pretty but it works :)
Thanks a lot!
